decompiler/pipeline/controlflowanalysis/expression_simplification_rules.py

```diff
@@ -0,0 +1,145 @@
+    SimplifyTrivialBitArithmetic,
+)
+from decompiler.pipeline.controlflowanalysis.expression_simplification.rules.simplify_trivial_logic_arithmetic import (
+    SimplifyTrivialLogicArithmetic,
+)
+from decompiler.pipeline.controlflowanalysis.expression_simplification.rules.simplify_trivial_shift import SimplifyTrivialShift
+from decompiler.pipeline.controlflowanalysis.expression_simplification.rules.sub_to_add import SubToAdd
+from decompiler.pipeline.controlflowanalysis.expression_simplification.rules.term_order import TermOrder
+from decompiler.pipeline.stage import PipelineStage
+from decompiler.structures.ast.ast_nodes import CodeNode
+from decompiler.structures.pseudo import Instruction, Operation
+from decompiler.structures.visitors.substitute_visitor import SubstituteVisitor
+from decompiler.task import DecompilerTask
+
+
+class _ExpressionSimplificationRulesBase(PipelineStage, ABC):
+
+    def run(self, task: DecompilerTask):
+        max_iterations = task.options.getint("expression-simplification.max_iterations")
+        simplify_instructions(self._get_instructions(task), max_iterations)
+
+    @abstractmethod
+    def _get_instructions(self, task: DecompilerTask) -> list[Instruction]:
+        pass
+
+
+class ExpressionSimplificationRulesCfg(_ExpressionSimplificationRulesBase):
+    """
+    Pipeline stage that simplifies cfg expressions by applying a set of simplification rules.
+    """
+
+    name = "expression-simplification-rules-cfg"
+
+    def _get_instructions(self, task: DecompilerTask) -> list[Instruction]:
+        return list(task.graph.instructions)
+
+
+class ExpressionSimplificationRulesAst(_ExpressionSimplificationRulesBase):
+    """
+    Pipeline stage that simplifies ast expressions by applying a set of simplification rules.
+    """
+
+    name = "expression-simplification-rules-ast"
+
+    def _get_instructions(self, task: DecompilerTask) -> list[Instruction]:
+        instructions = []
+        for node in task.syntax_tree.topological_order():
+            if isinstance(node, CodeNode):
+                instructions.extend(node.instructions)
+
+        return instructions
+
+
+_pre_rules: list[SimplificationRule] = []
+_rules: list[SimplificationRule] = [
+    TermOrder(),
+    SubToAdd(),
+    SimplifyRedundantReference(),
+    SimplifyTrivialArithmetic(),
+    SimplifyTrivialBitArithmetic(),
+    SimplifyTrivialLogicArithmetic(),
+    SimplifyTrivialShift(),
+    CollapseConstants(),
+    CollectTerms(),
+]
+_post_rules: list[SimplificationRule] = [
+    CollapseAddNeg(),
+    FixAddSubSign()
+]
+
+
+def simplify_instructions(instructions: list[Instruction], max_iterations: int):
+    rule_sets = [
+        ("pre-rules", _pre_rules),
+        ("rules", _rules),
+        ("post-rules", _post_rules)
+    ]
+    for rule_name, rule_set in rule_sets:
+        iteration_count = _simplify_instructions_with_rule_set(instructions, rule_set, max_iterations)
```

Comment:
My idea behind design it like this was, that when when one rule set fails because of to many iterations, `post-rules`, which is largely just cleaning up stuff, can still run.
I've slightly changed the documentation in default.json with 69060a95, however I'm not sure how detailed the explanation should be.